controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -150,6 +154,11 @@ func (r *KubeadmControlPlaneReconciler) Reconcile(req ctrl.Request) (res ctrl.Re
 		return ctrl.Result{}, nil
 	}
 
+	// Make sure to reconcile the external infrastructure reference.
+	if err := r.reconcileExternalReference(ctx, cluster, kcp.Spec.InfrastructureTemplate); err != nil {
+		return ctrl.Result{RequeueAfter: DependentInfraTemplateRequeueAfter}, err
```

Comment:
If the error is not nil, the reconciler will go in exponential backoff. What's the reasoning around moving the call outside of the `reconcile` method?

Reply:
Here we will get following errors:

- InfrastructureTemplate is not found, will be created in the future. 
- failed to set ownerReference to InfrastructureTemplate
- connection issue 

under such situations, we should retry reconciling InfrastructureTemplate.

moving the call outside of the reconcile because when InfrastructureTemplate doesn't exits.

Reply:
That should be ok, the `err` is returned and the reconciler will go in exponential backoff. From the trace posted in your issue, it seems like this should be fixed in https://github.com/kubernetes-sigs/cluster-api/issues/2754